File: PalaceCardGame/Player.cpp
//Player.cpp
// Created by Craig Gleckman
#include "Player.h"
#include <iostream>

Player::Player()
{
	//Initialize private variables
	//Create control Card for monitoring
	Card controlCard;
	controlCard.setRank(999);
	controlCard.setSuit('U');

	playerNum = -1337;
	name = "NOT SET";

	palace, topOfPalace, hand;
	
	/* KEEP UNTIL I know i dont need to revert palace and topOfPalace to arrays
	//Initialize palace and topOfPalace arrays
	//Intial cards are all control cards for monitoring
	palace[0].push_back(controlCard);
	palace[1] = controlCard;
	palace[2] = controlCard;
	topOfPalace[0] = controlCard;
	topOfPalace[1] = controlCard;
	topOfPalace[2] = controlCard;*/
}

Player::~Player()
{
}

int Player::getPlayerNum()
{
	return playerNum;
}

void Player::setPlayerNum(const int PLAYERNUM)
{
	this->playerNum = PLAYERNUM;
}

string Player::getName()
{
	return name;
}

void Player::setName(string name)
{
	this->name = name;
}

int Player::getPalaceCount()
{
	cout << "Palace Count: " << palace.size() << endl;
	return palace.size();
}

Card Player::getPalaceCard(int pos)
{
	return palace[pos];
}

void Player::setPalaceCard(Card card)
{
	if(palace.size() > 3)
	{
		cout << "SOMEHOW IM TO HIGH" << endl;
	}
	else
		palace.push_back(card);
}

int Player::getTopOfPalaceCount()
{
	return topOfPalace.size();
}

Card Player::getTopOfPalaceCard(int pos)
{
	return topOfPalace[pos];
}

vector<Card> Player::getTopOfPalace()
{
	return topOfPalace;
}

void Player::setTopOfPalace(const Card CARD)
{
	topOfPalace.push_back(CARD);
}

vector<Card> Player::getHand()
{
	return hand;
}

void Player::setHand(vector<Card> &cards)
{
	hand = cards;
}

int Player::getHandCount()
{
	return hand.size();
}

Card Player::getHandCard(int pos)
{	
	return hand[pos];
}

Card Player::getHighestHandCard()
{
	Card highestCard = hand[0];
	for(int i=1; i<=hand.size(); i++)
	{
		if(hand[i].getRank() > highestCard.getRank())
		{
			highestCard = hand[i];
		}
	}
	return highestCard;
}

void Player::setHandCard(Card card)
{
	hand.push_back(card);
}

void Player::removeHandCard(int pos)
{

	hand[pos].DisplayCard();
	cout << endl;
	if(pos == hand.size()-1)
	{
		hand.pop_back();
		hand[pos-1].DisplayCard();
		cout << endl;
	}
	else if(pos < hand.size()-1)
	{
		hand.erase(hand.begin()+pos);
		cout << "After Remove Card: ";
		hand[pos].DisplayCard(); 
		cout << endl;
	}
	else
	{
		cout << "Something went wrong IN removeHandCard ** ERASE DETERMINATION ** " << endl;
	}
}

void Player::DisplayHand()
{
	cout << "Current Hand: " << endl;

	for(int i=0; i<hand.size(); i++)
	{
		cout << i << ". " << hand[i].getRank() 
			<< " " << hand[i].getSuit() << endl;
	}
}

File: PalaceCardGamePythonWebApp/Player.py
#-------------------------------------------------------------------------------
# Name:        Player.py
# Purpose:
#
# Author:      Craig Gleckman
#
# Created:     25/03/2014
# Copyright:   (c) Craig 2014
# Licence:     <your licence>
#-------------------------------------------------------------------------------

#Questions for Charlie:
#   1. How to handle the need to make sure things are cards?
#       a. make hand/palace/topofpalace into own objects
#       b. Put checks in setters (and getters for double checking?)
#       c. ????
#   2. Can I grab the assert if it fails and do something with it? EX: if assert false do something?
from Card import Card

class Player(object):

    def __init__(self):
        #Initialize private variables
        #Create control Card for monitoring
        self.controlCard = Card()
        self.controlCard.setRank(999)
        self.controlCard.setSuit('U')
        self.CONST_PLAYERNUM = -1337
        self.name = "NOT SET"
        self.palace = []
        self.topOfPalace = []
        #code: type(var) -- will give type of whatever is between ()
        #code: assert type(var) == type(Card())
        self.hand = []

    def getPlayerNum(self):
        return self.playerNum

    def setPlayerNum(self, playerNum):
        if self.CONST_PLAYERNUM == -1337:
            assert type(playerNum) == type(int())
            self.CONST_PLAYERNUM = playerNum
        else:
            print "Error in setPlayerNum: Attempting to change player number after initial set"

    def getName(self):
        if self.name != "NOT SET":
            return self.name
        else:
            print "Error in getName: Name has not yet been set"

    def setName(self, name):
        if self.name == "NOT SET":
            #add while loop till passes assert
            try:
                print "Name type: ", type(name), type(str())
                assert type(name) == type(str())
            except AssertionError:
                #determine type and if callable turn to string
                print "Assert failed in setName"
            self.name = name
        else:
            print "Error in setName: Name has already been set"

    def getPalaceCount(self):
        print "Palace Count: %s" % (len(self.palace))
        return len(self.palace)

    def getPalaceCard(self, pos):
        assert type(pos) == type(int())
        if type(self.palace[pos]) == type(Card()):
            return self.palace[pos]
        #add further functionality to deal with errors: Convert to Card if possible?
        else:
            print "Error in getPalaceCard: Item stored in palace is not a CARD"

    def setPalaceCard(self, card):
        assert type(card) == type(Card())
        if len(self.palace) < 4:
            self.palace.append(card)
            print "Appending Card: ", card.getRank(), card.getSuit(), "Turn: ", self.getName()
        else:
            print "Error in setPalaceCard: To to many/little amount of cards in Palace"

    def getTopOfPalaceCount(self):
        return len(self.topOfPalace)

    def getTopOfPalaceCard(self, pos):
        assert type(pos) == type(int())
        assert type(self.topOfPalace[pos]) == type(Card())
        return self.topOfPalace[pos]

    def getTopOfPalace(self):
        return self.topOfPalace

    def setTopOfPalace(self, CONST_CARD):
        assert type(CONST_CARD) == type(Card())
        self.topOfPalace.append(CONST_CARD)

    def getHand(self):
        return self.hand

    def setHand(self, cards):
        for card in cards:
            assert type(card) == type(Card())
            self.hand.append(cards)

    def getHandCount(self):
        return len(self.hand)

    def getHandCard(self, pos):
        assert type(pos) == type(int())
        return self.hand[pos]

    #throws error: TypeError: 'list' object is not callable --
    #appears at if Statement
    def getHighestHandCard(self):
        highestCard = self.hand[0]
        print highestCard
        for pos in range(len(self.hand)):
            if self.hand[pos].getRank() > highestCard.getRank():
                highestCard = self.hand(card)
                return highestCard

    def setHandCard(self, card):
        assert type(card) == type(Card())
        self.hand.append(card)

    def removeHandCard(self, pos):
        assert type(pos) == type(int())
        self.hand[pos].DisplayCard()
        if pos == len(self.hand) - 1:
            self.hand.pop()
        #requires testing to make sure pop is done right (may not remove intended card)
        elif pos < len(self.hand) - 1:
            self.hand.pop(pos)
            print "After Remove Card: %s" % (self.hand[pos].DisplayCard())
        else:
            print "Something went wrong IN removeHandCard ** ERASE DETERMINATION ** "
    #same error as above: TypeError: 'list' object is not callable -- appears at print
    def DisplayHand(self):
        print "Current Hand: "
        num = -1
        for card in self.hand:
            num += 1
            print "%s. %s" % (num, card.DisplayCard())
